### ISSUE TITLE
windows x64 uses SOCKET (64bit) instead of int (32 bit) as descriptor. BIO_new_socket takes int

### ISSUE THREAD

**Issue by leleftheriades** [openssl/openssl#8169]
In windows the descriptor for socket is SOCKET rather than int.
When building in x64, SOCKET is 64 bits. I see that the following methods assume that a socket is int:

BIO_new_socket
BIO_connect
BIO_listen
BIO_accept_ex
BIO_closesocket
BIO_sock_error
BIO_sock_info

Is it possible to add support for x64 windows sockets?

**Comment by kaduk** [comment]
@leleftheriades could you say something about why you closed the issue?

**Comment by leleftheriades** [comment]
sorry, yes, I closed it because I realized that the authors were aware of this issue yet rationalized the current implementation. In include/internal/sockets.h I found the following comment:

Even though sizeof(SOCKET) is 8, it's safe to cast it to int, because
the value constitutes an index in per-process table of limited size
and not a real pointer. And we also depend on fact that all processors
Windows run on happen to be two's-complement, which allows to
interchange INVALID_SOCKET and -1.

While by no means this rationalization makes me happy, I thought it pointless to open an issue on something the authors were aware.

I examined bio_ssl.c, bio_sock.c, bio.h, bio_lib.c to figure out if there is an easy solution but i got ping-ponged between these files unable to figure out the flow
